File: agentprimordia/internal/admin/handler.go
package admin

import (
	"encoding/json"
	"html"
	"log/slog"
	"net/http"
	"runtime"
	"time"

	"agentprimordia/internal/pool"
	"agentprimordia/internal/tools"
)

type AdminHandler struct {
	pool     *pool.Pool
	registry *tools.Registry
	mux      *http.ServeMux
	logger   *slog.Logger
}

func NewAdminHandler(p *pool.Pool, r *tools.Registry) *AdminHandler {
	h := &AdminHandler{
		pool:     p,
		registry: r,
		mux:      http.NewServeMux(),
		logger:   slog.Default(),
	}

	// 原有端点
	h.mux.HandleFunc("GET /api/agents", h.listAgents)
	h.mux.HandleFunc("GET /api/agents/{id}", h.getAgent)
	h.mux.HandleFunc("GET /api/stats", h.stats)
	h.mux.HandleFunc("GET /api/tasks", h.tasks)
	h.mux.HandleFunc("GET /api/health", h.health)
	h.mux.HandleFunc("GET /api/system", h.systemInfo)

	// 新增工具管理端点
	h.mux.HandleFunc("GET /api/tools", h.listTools)
	h.mux.HandleFunc("GET /api/tools/{name}", h.getTool)
	h.mux.HandleFunc("GET /api/tools/categories", h.toolCategories)

	// 新增工作流监控端点
	h.mux.HandleFunc("GET /api/workflows", h.listWorkflows)
	h.mux.HandleFunc("GET /api/workflows/{id}", h.getWorkflow)

	// 新增实时日志端点
	h.mux.HandleFunc("GET /api/logs/stream", h.logStream)

	h.mux.HandleFunc("GET /", h.index)

	return h
}

func (h *AdminHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	h.mux.ServeHTTP(w, r)
}

func (h *AdminHandler) listAgents(w http.ResponseWriter, r *http.Request) {
	agents := h.pool.ListAgents()
	writeJSON(w, http.StatusOK, agents)
}

func (h *AdminHandler) getAgent(w http.ResponseWriter, r *http.Request) {
	id := r.PathValue("id")
	if id == "" {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "缺少 agent ID"})
		return
	}

	agents := h.pool.ListAgents()
	if status, exists := agents[id]; exists {
		writeJSON(w, http.StatusOK, map[string]any{
			"id":     id,
			"status": status,
		})
		return
	}

	writeJSON(w, http.StatusNotFound, map[string]string{"error": "agent 未找到"})
}

func (h *AdminHandler) stats(w http.ResponseWriter, r *http.Request) {
	stats := h.pool.Stats()
	writeJSON(w, http.StatusOK, stats)
}

func (h *AdminHandler) tasks(w http.ResponseWriter, r *http.Request) {
	tasks := h.pool.ListTasks()

	items := make([]map[string]any, 0, len(tasks))
	for _, t := range tasks {
		items = append(items, taskResultToJSON(t))
	}

	writeJSON(w, http.StatusOK, items)
}

func (h *AdminHandler) health(w http.ResponseWriter, r *http.Request) {
	stats := h.pool.Stats()
	writeJSON(w, http.StatusOK, map[string]any{
		"status":    "healthy",
		"timestamp": time.Now().UTC().Format(time.RFC3339),
		"tasks":     stats.TotalTasks,
		"running":   stats.RunningTasks,
	})
}

func (h *AdminHandler) systemInfo(w http.ResponseWriter, r *http.Request) {
	var memStats runtime.MemStats
	runtime.ReadMemStats(&memStats)

	writeJSON(w, http.StatusOK, map[string]any{
		"go_version":   runtime.Version(),
		"goroutines":   runtime.NumGoroutine(),
		"cpu_count":    runtime.NumCPU(),
		"mem_alloc_mb": float64(memStats.Alloc) / 1024 / 1024,
		"mem_sys_mb":   float64(memStats.Sys) / 1024 / 1024,
		"gc_count":     memStats.NumGC,
		"heap_objects": memStats.HeapObjects,
		"stack_use_mb": float64(memStats.StackInuse) / 1024 / 1024,
	})
}

// listTools 列出所有已注册工具
func (h *AdminHandler) listTools(w http.ResponseWriter, r *http.Request) {
	if h.registry == nil {
		writeJSON(w, http.StatusOK, []any{})
		return
	}
	toolsList := h.registry.List()
	result := make([]map[string]any, 0, len(toolsList))
	for _, name := range toolsList {
		t, ok := h.registry.Get(name)
		if !ok {
			continue
		}
		result = append(result, map[string]any{
			"name":        t.Name(),
			"description": t.Description(),
			"parameters":  t.Parameters(),
		})
	}
	writeJSON(w, http.StatusOK, result)
}

// getTool 获取单个工具详情
func (h *AdminHandler) getTool(w http.ResponseWriter, r *http.Request) {
	if h.registry == nil {
		writeJSON(w, http.StatusNotFound, map[string]string{"error": "工具注册表为空"})
		return
	}
	name := r.PathValue("name")
	if name == "" {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "缺少工具名称"})
		return
	}
	tool, ok := h.registry.Get(name)
	if !ok {
		writeJSON(w, http.StatusNotFound, map[string]string{"error": "工具不存在"})
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{
		"name":        tool.Name(),
		"description": tool.Description(),
		"parameters":  tool.Parameters(),
	})
}

// toolCategories 按类别列出工具
func (h *AdminHandler) toolCategories(w http.ResponseWriter, r *http.Request) {
	if h.registry == nil {
		writeJSON(w, http.StatusOK, map[string]any{})
		return
	}
	categories := h.registry.ToolsByCategory()
	writeJSON(w, http.StatusOK, categories)
}

// listWorkflows 列出所有工作流（基于 Pool 任务列表）
func (h *AdminHandler) listWorkflows(w http.ResponseWriter, r *http.Request) {
	if h.pool == nil {
		writeJSON(w, http.StatusOK, []any{})
		return
	}

	tasks := h.pool.ListTasks()
	result := make([]map[string]any, 0, len(tasks))
	for _, task := range tasks {
		entry := map[string]any{
			"id":       task.TaskID,
			"title":    task.Task.Title,
			"status":   string(task.Status),
			"duration": task.Duration.String(),
		}
		if task.Task.SessionID != "" {
			entry["session_id"] = task.Task.SessionID
		}
		if task.Error != nil {
			entry["error"] = task.Error.Error()
		}
		result = append(result, entry)
	}
	writeJSON(w, http.StatusOK, result)
}

// getWorkflow 获取单个工作流详情
func (h *AdminHandler) getWorkflow(w http.ResponseWriter, r *http.Request) {
	id := r.PathValue("id")
	if id == "" {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "缺少工作流 ID"})
		return
	}
	if h.pool == nil {
		writeJSON(w, http.StatusNotFound, map[string]string{"error": "工作流未找到"})
		return
	}

	task, ok := h.pool.GetTask(id)
	if !ok {
		writeJSON(w, http.StatusNotFound, map[string]string{"error": "工作流未找到"})
		return
	}

	result := map[string]any{
		"id":       task.TaskID,
		"title":    task.Task.Title,
		"prompt":   task.Task.Prompt,
		"status":   string(task.Status),
		"duration": task.Duration.String(),
	}
	if task.Task.SessionID != "" {
		result["session_id"] = task.Task.SessionID
	}
	if task.Response != nil {
		result["response"] = task.Response.Content
	}
	if task.Error != nil {
		result["error"] = task.Error.Error()
	}
	writeJSON(w, http.StatusOK, result)
}

// logStream 实时日志流（Server-Sent Events）
func (h *AdminHandler) logStream(w http.ResponseWriter, r *http.Request) {
	flusher, ok := w.(http.Flusher)
	if !ok {
		http.Error(w, "Streaming not supported", http.StatusInternalServerError)
		return
	}

	w.Header().Set("Content-Type", "text/event-stream")
	w.Header().Set("Cache-Control", "no-cache")
	w.Header().Set("Connection", "keep-alive")

	// 发送初始连接确认
	_, _ = w.Write([]byte("event: connected\ndata: {\"status\":\"ok\"}\n\n"))
	flusher.Flush()

	// 保持连接直到客户端断开
	<-r.Context().Done()
}

func (h *AdminHandler) index(w http.ResponseWriter, r *http.Request) {
	if r.URL.Path != "/" {
		writeJSON(w, http.StatusNotFound, map[string]string{"error": "未找到"})
		return
	}

	w.Header().Set("Content-Type", "text/html; charset=utf-8")
	_, _ = w.Write([]byte(indexHTML))
}

func writeJSON(w http.ResponseWriter, status int, v any) {
	w.Header().Set("Content-Type", "application/json; charset=utf-8")
	w.WriteHeader(status)
	_ = json.NewEncoder(w).Encode(v)
}

// taskResultToJSON 将 TaskResult 转为可 JSON 序列化的 map
// 处理 error 和 time.Duration 的序列化问题
func taskResultToJSON(t pool.TaskResult) map[string]any {
	m := map[string]any{
		"task_id": t.TaskID,
		"task":    t.Task,
		"status":  t.Status,
	}
	if t.Error != nil {
		m["error"] = t.Error.Error()
	}
	if t.Duration > 0 {
		m["duration_ms"] = t.Duration.Milliseconds()
	}
	if t.Response != nil {
		m["response"] = t.Response
	}
	return m
}

// sanitizeHTML 对字符串进行 HTML 转义，防止 XSS
func sanitizeHTML(s string) string {
	return html.EscapeString(s)
}

const indexHTML = `<!DOCTYPE html>
<html lang="zh-CN">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width,initial-scale=1">
<title>AgentPrimordia 管理面板</title>
<style>
*{margin:0;padding:0;box-sizing:border-box}
body{font-family:system-ui,-apple-system,sans-serif;background:#0f172a;color:#e2e8f0;padding:24px}
h1{font-size:1.5rem;margin-bottom:16px;color:#38bdf8}
.stats{display:grid;grid-template-columns:repeat(auto-fit,minmax(140px,1fr));gap:12px;margin-bottom:24px}
.stat{background:#1e293b;border-radius:8px;padding:16px;text-align:center}
.stat .num{font-size:1.8rem;font-weight:700;color:#38bdf8}
.stat .lbl{font-size:.8rem;color:#94a3b8;margin-top:4px}
table{width:100%;border-collapse:collapse;background:#1e293b;border-radius:8px;overflow:hidden}
th,td{padding:10px 14px;text-align:left;border-bottom:1px solid #334155}
th{background:#0f172a;color:#94a3b8;font-size:.8rem;text-transform:uppercase}
td{font-size:.9rem}
.tag{display:inline-block;padding:2px 8px;border-radius:4px;font-size:.75rem;font-weight:600}
.tag-queued{background:#fbbf24;color:#1e293b}
.tag-running{background:#38bdf8;color:#1e293b}
.tag-completed{background:#4ade80;color:#1e293b}
.tag-failed{background:#f87171;color:#1e293b}
.tag-cancelled{background:#94a3b8;color:#1e293b}
.refresh{font-size:.75rem;color:#64748b;margin-bottom:12px}
</style>
</head>
<body>
<h1>AgentPrimordia 管理面板</h1>
<div class="refresh">每 5 秒自动刷新</div>
<div id="stats" class="stats"></div>
<table>
<thead><tr><th>ID</th><th>标题</th><th>状态</th><th>耗时(ms)</th><th>错误</th></tr></thead>
<tbody id="tasks"></tbody>
</table>
<script>
function tagClass(s){return 'tag tag-'+s}
function esc(s){if(s==null)return '';var d=document.createElement('div');d.textContent=s;return d.innerHTML}
function fetchAndUpdate(){
  Promise.all([
    fetch('/api/stats').then(r=>r.json()),
    fetch('/api/tasks').then(r=>r.json())
  ]).then(([stats,tasks])=>{
    document.getElementById('stats').innerHTML=
      '<div class="stat"><div class="num">'+stats.total_tasks+'</div><div class="lbl">总任务</div></div>'+
      '<div class="stat"><div class="num">'+stats.running_tasks+'</div><div class="lbl">运行中</div></div>'+
      '<div class="stat"><div class="num">'+stats.completed_tasks+'</div><div class="lbl">已完成</div></div>'+
      '<div class="stat"><div class="num">'+stats.failed_tasks+'</div><div class="lbl">失败</div></div>';
    var rows='';
    for(var i=0;i<tasks.length;i++){
      var t=tasks[i];
      var dur=t.duration_ms!=null?t.duration_ms:'-';
      var err=t.error||'';
      rows+='<tr><td>'+esc(t.task_id)+'</td><td>'+esc(t.task&&t.task.title||'-')+'</td>'+
        '<td><span class="'+tagClass(t.status)+'">'+esc(t.status)+'</span></td>'+
        '<td>'+dur+'</td><td>'+esc(err)+'</td></tr>';
    }
    document.getElementById('tasks').innerHTML=rows;
  }).catch(()=>{});
}
fetchAndUpdate();
setInterval(fetchAndUpdate,5000);
</script>
</body>
</html>`

File: sdk/typescript/src/admin/handler.ts
// Admin HTTP API — Bearer Token authenticated management endpoints
// Mirrors Go internal/admin/handler.go

import type { IncomingMessage, ServerResponse } from 'node:http';
import type { ToolRegistry } from '../tools/registry.js';
import type { AgentPool, PoolTask, PoolResult } from '../pool/agent-pool.js';

// ===== Admin Handler Config =====

export interface AdminHandlerConfig {
  pool?: AgentPool;
  registry?: ToolRegistry;
  apiToken: string;
  /** Optional: provide task history for /api/tasks and /api/workflows */
  getTasks?: () => AdminTaskInfo[];
  /** Optional: provide agent list for /api/agents */
  getAgents?: () => Record<string, string>;
  /** Optional: provide pool stats for /api/stats */
  getStats?: () => AdminPoolStats;
}

export interface AdminTaskInfo {
  id: string;
  title: string;
  status: string;
  duration: string;
  sessionId?: string;
  error?: string;
  prompt?: string;
  response?: string;
}

export interface AdminPoolStats {
  totalTasks: number;
  completedTasks: number;
  failedTasks: number;
  runningTasks: number;
  queuedTasks: number;
  maxConcurrency: number;
  activeConcurrency: number;
}

// ===== Admin Handler =====

export class AdminHandler {
  private config: AdminHandlerConfig;
  private routes: Map<string, (req: IncomingMessage, res: ServerResponse) => void | Promise<void>>;

  constructor(config: AdminHandlerConfig) {
    this.config = config;
    this.routes = new Map();

    // Public health endpoint
    this.routes.set('GET /api/health', this.health.bind(this));

    // Protected endpoints
    this.routes.set('GET /api/agents', this.requireAuth(this.listAgents.bind(this)));
    this.routes.set('GET /api/agents/', this.requireAuth(this.getAgent.bind(this)));
    this.routes.set('GET /api/stats', this.requireAuth(this.stats.bind(this)));
    this.routes.set('GET /api/tasks', this.requireAuth(this.tasks.bind(this)));
    this.routes.set('GET /api/system', this.requireAuth(this.systemInfo.bind(this)));
    this.routes.set('GET /api/tools', this.requireAuth(this.listTools.bind(this)));
    this.routes.set('GET /api/tools/', this.requireAuth(this.getTool.bind(this)));
    this.routes.set('GET /api/tools/categories', this.requireAuth(this.toolCategories.bind(this)));
    this.routes.set('GET /api/workflows', this.requireAuth(this.listWorkflows.bind(this)));
    this.routes.set('GET /api/workflows/', this.requireAuth(this.getWorkflow.bind(this)));
    this.routes.set('GET /api/logs/stream', this.requireAuth(this.logStream.bind(this)));
    this.routes.set('GET /', this.index.bind(this));
  }

  /** Handle an HTTP request. */
  async handle(req: IncomingMessage, res: ServerResponse): Promise<void> {
    const method = req.method ?? 'GET';
    const url = req.url ?? '/';

    // Try exact match first
    let handler = this.routes.get(`${method} ${url}`);

    // Try prefix match for path parameters
    if (!handler) {
      for (const [pattern, h] of this.routes) {
        const [pMethod, pPath] = pattern.split(' ');
        if (pMethod === method && pPath!.endsWith('/')) {
          if (url.startsWith(pPath!)) {
            handler = h;
            break;
          }
        }
      }
    }

    if (!handler) {
      this.writeJSON(res, 404, { error: 'Not found' });
      return;
    }

    try {
      await handler(req, res);
    } catch (err) {
      this.writeJSON(res, 500, { error: (err as Error).message });
    }
  }

  /** Auth middleware wrapper. */
  private requireAuth(
    next: (req: IncomingMessage, res: ServerResponse) => void | Promise<void>
  ): (req: IncomingMessage, res: ServerResponse) => void | Promise<void> {
    return (req, res) => {
      if (!this.config.apiToken) {
        this.writeJSON(res, 401, { error: 'Admin API token not configured' });
        return;
      }

      const auth = req.headers['authorization'] ?? '';
      const prefix = 'Bearer ';
      if (!auth.startsWith(prefix) || auth.slice(prefix.length) !== this.config.apiToken) {
        this.writeJSON(res, 401, { error: 'Invalid or missing Bearer token' });
        return;
      }

      return next(req, res);
    };
  }

  // ===== Endpoint Handlers =====

  private health(_req: IncomingMessage, res: ServerResponse): void {
    const stats = this.config.getStats?.() ?? { totalTasks: 0, runningTasks: 0 } as Partial<AdminPoolStats>;
    this.writeJSON(res, 200, {
      status: 'healthy',
      timestamp: new Date().toISOString(),
      tasks: stats.totalTasks ?? 0,
      running: stats.runningTasks ?? 0,
    });
  }

  private listAgents(_req: IncomingMessage, res: ServerResponse): void {
    const agents = this.config.getAgents?.() ?? {};
    this.writeJSON(res, 200, agents);
  }

  private getAgent(req: IncomingMessage, res: ServerResponse): void {
    const id = (req.url ?? '').split('/').pop() ?? '';
    if (!id) {
      this.writeJSON(res, 400, { error: 'Missing agent ID' });
      return;
    }

    const agents = this.config.getAgents?.() ?? {};
    if (id in agents) {
      this.writeJSON(res, 200, { id, status: agents[id] });
    } else {
      this.writeJSON(res, 404, { error: 'Agent not found' });
    }
  }

  private stats(_req: IncomingMessage, res: ServerResponse): void {
    const stats = this.config.getStats?.() ?? {
      totalTasks: 0, completedTasks: 0, failedTasks: 0,
      runningTasks: 0, queuedTasks: 0, maxConcurrency: 0, activeConcurrency: 0,
    };
    this.writeJSON(res, 200, stats);
  }

  private tasks(_req: IncomingMessage, res: ServerResponse): void {
    const tasks = this.config.getTasks?.() ?? [];
    this.writeJSON(res, 200, tasks);
  }

  private systemInfo(_req: IncomingMessage, res: ServerResponse): void {
    const mem = process.memoryUsage();
    this.writeJSON(res, 200, {
      node_version: process.version,
      platform: process.platform,
      arch: process.arch,
      pid: process.pid,
      uptime_seconds: Math.floor(process.uptime()),
      mem_rss_mb: Math.round((mem.rss / 1024 / 1024) * 100) / 100,
      mem_heap_used_mb: Math.round((mem.heapUsed / 1024 / 1024) * 100) / 100,
      mem_heap_total_mb: Math.round((mem.heapTotal / 1024 / 1024) * 100) / 100,
    });
  }

  private listTools(_req: IncomingMessage, res: ServerResponse): void {
    if (!this.config.registry) {
      this.writeJSON(res, 200, []);
      return;
    }
    const tools = this.config.registry.list().map((t) => ({
      name: t.name,
      description: t.description,
      parameters: t.parameters,
    }));
    this.writeJSON(res, 200, tools);
  }

  private getTool(req: IncomingMessage, res: ServerResponse): void {
    if (!this.config.registry) {
      this.writeJSON(res, 404, { error: 'Tool registry is empty' });
      return;
    }
    const name = (req.url ?? '').split('/').pop() ?? '';
    if (!name) {
      this.writeJSON(res, 400, { error: 'Missing tool name' });
      return;
    }
    const tool = this.config.registry.get(name);
    if (!tool) {
      this.writeJSON(res, 404, { error: 'Tool not found' });
      return;
    }
    this.writeJSON(res, 200, {
      name: tool.name,
      description: tool.description,
      parameters: tool.parameters,
    });
  }

  private toolCategories(_req: IncomingMessage, res: ServerResponse): void {
    if (!this.config.registry) {
      this.writeJSON(res, 200, {});
      return;
    }
    // Group tools by name prefix or "general"
    const categories: Record<string, string[]> = {};
    for (const tool of this.config.registry.list()) {
      const cat = tool.name.includes('_') ? tool.name.split('_')[0]! : 'general';
      if (!categories[cat]) categories[cat] = [];
      categories[cat].push(tool.name);
    }
    this.writeJSON(res, 200, categories);
  }

  private listWorkflows(_req: IncomingMessage, res: ServerResponse): void {
    const tasks = this.config.getTasks?.() ?? [];
    this.writeJSON(res, 200, tasks.map((t) => ({
      id: t.id,
      title: t.title,
      status: t.status,
      duration: t.duration,
      session_id: t.sessionId,
      error: t.error,
    })));
  }

  private getWorkflow(req: IncomingMessage, res: ServerResponse): void {
    const id = (req.url ?? '').split('/').pop() ?? '';
    if (!id) {
      this.writeJSON(res, 400, { error: 'Missing workflow ID' });
      return;
    }
    const tasks = this.config.getTasks?.() ?? [];
    const task = tasks.find((t) => t.id === id);
    if (!task) {
      this.writeJSON(res, 404, { error: 'Workflow not found' });
      return;
    }
    this.writeJSON(res, 200, task);
  }

  private logStream(req: IncomingMessage, res: ServerResponse): void {
    res.writeHead(200, {
      'Content-Type': 'text/event-stream',
      'Cache-Control': 'no-cache',
      'Connection': 'keep-alive',
    });
    res.write('data: {"message":"Log stream connected"}\n\n');
    // Keep connection alive with heartbeat
    const interval = setInterval(() => {
      res.write(': heartbeat\n\n');
    }, 30000);
    req.on('close', () => clearInterval(interval));
  }

  private index(_req: IncomingMessage, res: ServerResponse): void {
    res.writeHead(200, { 'Content-Type': 'text/html; charset=utf-8' });
    res.end(ADMIN_HTML);
  }

  // ===== Helpers =====

  private writeJSON(res: ServerResponse, status: number, data: unknown): void {
    res.writeHead(status, { 'Content-Type': 'application/json; charset=utf-8' });
    res.end(JSON.stringify(data));
  }
}

const ADMIN_HTML = `<!DOCTYPE html>
<html>
<head>
  <title>AgentPrimordia Admin</title>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <style>
    body { font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, sans-serif; margin: 0; padding: 20px; background: #f5f5f5; }
    .container { max-width: 1200px; margin: 0 auto; }
    h1 { color: #333; }
    .card { background: white; border-radius: 8px; padding: 20px; margin-bottom: 20px; box-shadow: 0 2px 10px rgba(0,0,0,0.1); }
    table { width: 100%; border-collapse: collapse; }
    th, td { padding: 12px; text-align: left; border-bottom: 1px solid #eee; }
    th { background: #f9f9f9; color: #666; }
    .status-ok { color: #28a745; font-weight: bold; }
    .status-error { color: #dc3545; font-weight: bold; }
    .status-running { color: #007bff; font-weight: bold; }
  </style>
</head>
<body>
  <div class="container">
    <h1>AgentPrimordia Admin Dashboard</h1>
    <div class="card">
      <h2>Health</h2>
      <div id="health">Loading...</div>
    </div>
    <div class="card">
      <h2>Agents</h2>
      <div id="agents">Loading...</div>
    </div>
    <div class="card">
      <h2>Tasks</h2>
      <div id="tasks">Loading...</div>
    </div>
    <div class="card">
      <h2>Tools</h2>
      <div id="tools">Loading...</div>
    </div>
  </div>
  <script>
    const token = new URLSearchParams(location.search).get('token') || '';
    const headers = token ? { 'Authorization': 'Bearer ' + token } : {};
    async function fetchJSON(path) {
      const res = await fetch(path, { headers });
      return res.json();
    }
    async function loadAll() {
      try {
        const health = await fetchJSON('/api/health');
        document.getElementById('health').innerHTML = '<pre>' + JSON.stringify(health, null, 2) + '</pre>';
      } catch {}
      try {
        const agents = await fetchJSON('/api/agents');
        document.getElementById('agents').innerHTML = '<pre>' + JSON.stringify(agents, null, 2) + '</pre>';
      } catch {}
      try {
        const tasks = await fetchJSON('/api/tasks');
        document.getElementById('tasks').innerHTML = '<pre>' + JSON.stringify(tasks, null, 2) + '</pre>';
      } catch {}
      try {
        const tools = await fetchJSON('/api/tools');
        document.getElementById('tools').innerHTML = '<pre>' + JSON.stringify(tools, null, 2) + '</pre>';
      } catch {}
    }
    loadAll();
    setInterval(loadAll, 5000);
  </script>
</body>
</html>`;
